lib/ansible/modules/database/postgresql/postgresql_table.py

```diff
@@ -0,0 +1,648 @@
+    including = module.params["including"]
+    newname = module.params["rename"]
+    storage_params = module.params["storage_params"]
+    truncate = module.params["truncate"]
+    columns = module.params["columns"]
+    storage_params = module.params["storage_params"]
+    sslrootcert = module.params["ssl_rootcert"]
+    session_role = module.params["session_role"]
+
+    # Check mutual exclusive parameters:
+    if state == 'absent' and (truncate or newname or columns or tablespace or
+                              like or storage_params or unlogged or
+                              owner or including):
+        module.fail_json(msg="%s: state=absent is mutually exclusive with: "
+                             "truncate, rename, columns, tablespace, "
+                             "including, like, storage_params, unlogged, owner" % table)
+
+    if truncate and (newname or columns or like or unlogged or
+                     storage_params or owner or tablespace or including):
+        module.fail_json(msg="%s: truncate is mutually exclusive with: "
+                             "rename, columns, like, unlogged, including, "
+                             "storage_params, owner, tablespace" % table)
+
+    if newname and (columns or like or unlogged or
+                    storage_params or owner or tablespace or including):
+        module.fail_json(msg="%s: rename is mutually exclusive with: "
+                             "columns, like, unlogged, including, "
+                             "storage_params, owner, tablespace" % table)
+
+    if like and columns:
+        module.fail_json(msg="%s: like and columns params are mutually exclusive" % table)
+    if including and not like:
+        module.fail_json(msg="%s: including param needs like param specified" % table)
+
+    # To use defaults values, keyword arguments must be absent, so
+    # check which values are empty and don't include in the **kw
+    # dictionary
+    params_map = {
+        "login_host": "host",
+        "login_user": "user",
+        "login_password": "password",
+        "port": "port",
+        "db": "database",
+        "ssl_mode": "sslmode",
+        "ssl_rootcert": "sslrootcert"
+    }
+    kw = dict((params_map[k], v) for (k, v) in iteritems(module.params)
+              if k in params_map and v != "" and v is not None)
+
+    if not HAS_PSYCOPG2:
+        module.fail_json(msg="the python psycopg2 module is required")
+
+    # If a login_unix_socket is specified, incorporate it here.
+    is_localhost = "host" not in kw or kw["host"] is None or kw["host"] == "localhost"
+    if is_localhost and module.params["login_unix_socket"] != "":
+        kw["host"] = module.params["login_unix_socket"]
+
+    if psycopg2.__version__ < '2.4.3' and sslrootcert is not None:
+        module.fail_json(msg='psycopg2 must be at least 2.4.3 in order to user the ssl_rootcert parameter')
+
+    try:
+        db_connection = psycopg2.connect(**kw)
```

Comment:
In addition to @Glandos comment, I think you can use it like so (or see [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L451)):
```python
import ansible.module_utils.postgres as pgutils
...
    try:
        pgutils.ensure_libs(sslrootcert=module.params.get('ssl_rootcert'))
        db_connection = psycopg2.connect(**kw)
...
    except pgutils.LibraryError as e:
         module.fail_json(msg="unable to connect to database: {0}".format(to_native(e)), exception=traceback.format_exc())
    except TypeError as e:
        if 'sslrootcert' in e.args[0]:
            module.fail_json(msg='Postgresql server must be at least version 8.4 to support sslrootcert',
                exception=traceback.format_exc())
...
```

You could also add `exception=traceback.format_exc()`to your exceptions like [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L462)

Reply:
and then remove the parts Glandos point you.

Reply:
@tcraxs , thank you for this note!
I tried to change it.
However, look at lines 134, 422, 451 [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L451) (your first link)

1. postgresql_db.py imports psycopg2 in line 134
2. anyway, it checks that psycopg2 was successfully imported in line 422
3. ensure_libs checks again that psycopg2 was imported and its version in addition.

Anyway, we must import psycopg2 in the module file and if psycopg2.version < '2.4.3' and sslrootcert is not None: is a quite simple condition to import and invoke a special function ensure_libs to avoid it.

To be short, I suggest the current version of this module related with psycopg2 and its version because it's more efficient and simplier.

Reply:
I'm totally fine with your explanation and it was just a starting point by my side. And you are right, that the usage in your module in this case is simpler, so no change needed.